### PR TITLE
correcting the calss name and extending Calculator class

### PR DESCRIPTION
**Pull request by yaaHosein**
let us work together!

### CODE REVIEW

File: develpoedCalculator.js [hunk 1]
```diff
@@ -1 +1,3 @@
-class developedCalculator {};
\ No newline at end of file
+class DevelopedCalculator {
+    developedCalculator = new Calculator();
```
Comment (SMH110) [inline develpoedCalculator.js]: DevelopedCalculator needs to extend Calculator class not create an instance of Calculator class 

```js
class DevelopedCalculator extends Calculator {
}

```
Please read more about extending here
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/extends

Comment (SMH110) [reply develpoedCalculator.js]: When `DevelopedCalculator` extends `Calculator`, it inherits all the public methods and properties of Calculator class. this mean the instances of `DevelopedCalculator` have render method (they inherit the render method from the `base` class )

**Comment by SMH110** [comment]
Now, create a HTML element that can host our new calculator.
this this HTML element an id or class name so we can use it as selector.
After that create instance of `DevelopedCalculator ` class and pass the HTML element selector to class constructor (same as you did before)

then render the calculator using the `render` method.

File: develpoedCalculator.js [hunk 2]
```diff
@@ -1 +1,8 @@
-class DevelopedCalculator extends Calculator {};
\ No newline at end of file
+class DevelopedCalculator extends Calculator {
+    constructor(selector) {
+        this.hostElem1 = document.querySelector(selector);
+    }
+
+}
+let developedCalculator = new DevelopedCalculator("#hostElem1");
```
Comment (SMH110) [inline develpoedCalculator.js]: can you please point me where this selector  ("#hostElem1") is used in HTML code?

File: develpoedCalculator.js [hunk 3]
```diff
@@ -1 +1,8 @@
-class DevelopedCalculator extends Calculator {};
\ No newline at end of file
+class DevelopedCalculator extends Calculator {
+    constructor(selector) {
```
Comment (SMH110) [inline develpoedCalculator.js]: you don't need to add `constructor` to this class yet.
`DevelopedCalculator` class is passing the `selector` to its parent automatically

**Comment by SMH110** [comment]
what about the other comment?

File: develpoedCalculator.js (hunk 2 above)
Comment (yaaHosein) [reply develpoedCalculator.js]: sorry it was one in the html file 
but I remembered maybe you do not prefer using numbers 
so I changed it to _
but I forgot to change it in html file also

**Comment by SMH110** [comment]
do you have another calculator rendered in the UI?
can you send a screenshot?

**Comment by yaaHosein** [comment]
Maybe I did not understand you 
I have CalculatorLogic 

And since I removed the old calculator I do not see any calculator interface 
Thanks

**Comment by SMH110** [comment]
> Maybe I did not understand you
> I have CalculatorLogic
> 
> And since I removed the old calculator I do not see any calculator interface
> Thanks

I think you have a typo 

```html
 <script src="developedCalculator.js">
```

while the file name is 
`develpoedCalculator.js` 
you just need to be an extra careful with these long names

**Comment by SMH110** [comment]
is it working now?

**Comment by yaaHosein** [comment]
If you are asking about the calculator interface 
No not yet 
Still now I cannot see it

**Comment by SMH110** [comment]
> If you are asking about the calculator interface
> No not yet
> Still now I cannot see it

Are you sure that the file name and the script src are the same?

I want you to read my question carefully

**Comment by yaaHosein** [comment]
When I was near to the laptop I checked it and it was 
InshaAllah I will check it also later 

Bye the way I do not expect the calculator interface appears because I removed the hard codes in html file

**Comment by yaaHosein** [comment]
Unfortunately as you expected 
Yes they was not same 

After correcting the file name 
The calculator appeared 
I do not know how it is appearing without clear html codes !!!

**Comment by SMH110** [comment]
> I do not know how it is appearing without clear html codes !!!

Your comment is shocking!  
if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero

**Comment by SMH110** [comment]
Have you pushed the file name correction?

**Comment by yaaHosein** [comment]
> > I do not know how it is appearing without clear html codes !!!
> 
> Your comment is shocking!
> if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
> If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero

1- I am still remembering that you told me very long time before  when I created 
  <div id="hostElement"></div>
That the calculator or something similar ... will live in it 
and until we finished the second calculator implantation I did not saw you are asking me to put any codes inside it and still now it looks same :
  <div id="hostElement"></div>
============
2_ Then why you did not tell me previously to remove the wrapper div with its content before we start the third implantation??
=======
So I started to know that you moved the html files codes to the JS file, in the second implantation R? 

Thanks

**Comment by SMH110** [comment]
> > > I do not know how it is appearing without clear html codes !!!
> > 
> > 
> > Your comment is shocking!
> > if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
> > If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero
> 
> 1- I am still remembering that you told me very long time before when I created
> 
> That the calculator or something similar ... will live in it and until we finished the second calculator implantation I did not saw you are asking me to put any codes inside it and still now it looks same :
> ============ 2_ Then why you did not tell me previously to remove the wrapper div with its content before we start the third implantation?? ======= So I started to know that you moved the html files codes to the JS file, in the second implantation R?
> Thanks

I didn't understand your points.
I still don't understand why you said 
"I do not know how it is appearing without clear html codes !!!"

**Comment by yaaHosein** [comment]
> > > > I do not know how it is appearing without clear html codes !!!
> > > 
> > > 
> > > Your comment is shocking!
> > > if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
> > > If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero
> > 
> > 
> > 1- I am still remembering that you told me very long time before when I created
> > That the calculator or something similar ... will live in it and until we finished the second calculator implantation I did not saw you are asking me to put any codes inside it and still now it looks same :
> > ============ 2_ Then why you did not tell me previously to remove the wrapper div with its content before we start the third implantation?? ======= So I started to know that you moved the html files codes to the JS file, in the second implantation R?
> > Thanks
> 
> I didn't understand your points.
> I still don't understand why you said
> "I do not know how it is appearing without clear html codes !!!"

This comment was before thinking deeper 
I was thinking how the calculator will work without buttons elements 
...etc

But I remembered later that we still have the buttons but in the js file

**Comment by yaaHosein** [comment]
I do not know what is the answer of JavaScript on that id and class names which are used in the render method 
Refers to which elements?!
I used to see the elements first in the html file 
And in that file we add them 
Thanks

**Comment by SMH110** [comment]
> I do not know what is the answer of JavaScript on that id and class names which are used in the render method
> Refers to which elements?!
> I used to see the elements first in the html file
> And in that file we add them
> Thanks

Sorry you question is not clear to me

**Comment by SMH110** [comment]
> I do not know what is the answer of JavaScript on that id and class names which are used in the render method
> Refers to which elements?!
> I used to see the elements first in the html file
> And in that file we add them
> Thanks

I just opened the app in chrome with developer tools open on the elements tab 
image

here you can see your HTML code, nothing changed 
in previous pull request, you created the HTML structure using javascript instead of writing using HTML 
This question should be asked long time ago and I am really surprised why you're asking these question now?!
IF this does not make sense, let me know because there's no point to learn new stuff if you didn't understand the previous one

**Comment by yaaHosein** [comment]
Why you are saying that we used JavaScript to write this code
I remember you started by asking me to create a host in html 
Then in the render we used the id and class names which was in the html 
You never asked me to remove the wrapper div with it content in the previous before finishing the new implementation 

Always I was annoyed by seeing two calculators in UI

 Recently asked me to remove the hard cod
I was thinking we are using html file also 
Not only JavaScript 
Also we used especial codes for the id and classes in the render method 
If that implementation was pure JavaScript so why we are using these codes 

Thanks

**Comment by SMH110** [comment]
> Why you are saying that we used JavaScript to write this code
> I remember you started by asking me to create a host in html
> Then in the render we used the id and class names which was in the html
> You never asked me to remove the wrapper div with it content in the previous before finishing the new implementation
> 
> Always I was annoyed by seeing two calculators in UI
> 
> Recently asked me to remove the hard cod
> I was thinking we are using html file also
> Not only JavaScript
> Thanks

I asked you to create a host element so you can host the calculator in it.
I didn't say that we created the host element in js, I said we created the calculator HTML code in the JS in the render method.

we used css classes and id which are exist in the HTML code, this does not mean the HTML code was needed 
if you go back and read all the email and the answers you will see that I said to you from the begining we are trying to create the HTML ui of the calculator using javascript.
this is why we used the same classes and ids, because those classes has css rule in the `.css` file

the reason I didn't ask you remove the HTML code, because that was a prototype you needed to follow 
also, I kew that you didn't memorize the HTML code you so can create the same structure using js 

If I give you paper with some text on it and I ask you translate to English
you have two options:
1. memorise the whole text and then start transliting it without going back to the paper
2. or use the paper as reference 

I gave the second option, I kept the HTML code there so you can see how you should create the same structure using js 

now you tell me, were you able to create the content of the render method without having to look at the HTML code?

**Comment by yaaHosein** [comment]
Thanks a lot 
No I was not able 

But because we didn’t not remove the old one 
I thought we are building the new one using the old html 

And because we did not create a new html file as we create a new JavaScript 

Another question please 

If you I removed the wrapper div from the beginning 
How then I will use classList and id ... 

Thanks

**Comment by yaaHosein** [comment]
> > I do not know how it is appearing without clear html codes !!!
> 
> Your comment is shocking!
> if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
> If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero

Please would you tell me how it is rendered without html codes 
You just sent me yesterday the html codes are still now existing ( but hidden ) 

Thanks

**Comment by SMH110** [comment]
> > > I do not know how it is appearing without clear html codes !!!
> > 
> > 
> > Your comment is shocking!
> > if you understood what you did in the `render` method; then you would know how the Calculator rendered without HTML code.
> > If you didn't understand then we need to re-build the calculator from scratch and go over everything again from zero
> 
> Please would you tell me how it is rendered without html codes
> You just sent me yesterday the html codes are still now existing ( but hidden )
> 
> Thanks

Are you referring to how `Calcualtor` render method is working or `DevelopedCalculator ` ?

**Comment by yaaHosein** [comment]
Calculator render method

**Comment by SMH110** [comment]
> Calculator render method

Now, you are telling me that you didn't understand anything you did in the `render` method?

**Comment by SMH110** [comment]
This is is part of the render method 

image

on line 43, we have created a `div` element uisng `docuement.createElement` API which is allow you create HTML element in javascript 

then on line 44, we added a css class to the div we created 

then on line 45, we created another div, this div is used for showing the result in the calculator 
on line 46, we added "result" class name to this div 
on line 47 we added id to this div 
on line 49, we added a text value to this div (this.currentResult) 

then on line 51, we saved the result div as reference on the class

**Comment by SMH110** [comment]
then on line 52, we append the result div to the first div we created (the wrapper div)

then we create HTML code for the buttons (0 - 9)
and we add the buttons code to the wrapper 


this the end of render method
image

we append the warapper div to the host element 
the host element is added to the page via HTML code, while the wrapper div created using javascript and added to the host element using javascript 

note: the wrapper div contains buttons and result div

**Comment by yaaHosein** [comment]
So the button’s elements is a part of JavaScript file not html file 
Right?!

**Comment by SMH110** [comment]
> So the button’s element are a part of JavaScript file not html file
> Right?!

yes, javascript can inject HTML code into the page

**Comment by yaaHosein** [comment]
Ok thanks

**Comment by SMH110** [comment]
> Ok thanks

did you understand now?
because you used say ok thanks previosuly?

**Comment by yaaHosein** [comment]
Not everything yet 
If I wanted to re build the render method now 
How I can add the class names and id name 
I do not have the wrapper div with its content 
Thanks

**Comment by SMH110** [comment]
> Not everything yet
> If I wanted to re build the render method now
> How I can add the class names and id name
> I do not have the wrapper div with its content
> Thanks

What you didn't understand?
and which class you're referring to?

**Comment by yaaHosein** [comment]
I will let you know what :

In the line 44, 46 and 47 

From where I can bring these class and id names?!

**Comment by SMH110** [comment]
> I will let you know what :
> 
> In the line 46 and 47
> 
> From where I can bring these class and id names?!

have a look at the `calculator interface.css` file

**Comment by yaaHosein** [comment]
I did

**Comment by yaaHosein** [comment]
so you mean we can not start this implementation without having the class and id names from before 
Right?!

**Comment by yaaHosein** [comment]
I have replaced all resultElement with resultDiv

the calculator is working
but I got this error :

CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
    at Calculator.render (CalcualtorLogic.js:101)
    at CalcualtorLogic.js:191
render @ CalcualtorLogic.js:101
(anonymous) @ CalcualtorLogic.js:191

**Comment by SMH110** [comment]
> so you mean we can not start this implementation without having the class and id names from before
> Right?!

I didn't understand your question

**Comment by SMH110** [comment]
> I have replaced all resultElement with resultDiv
> 
> the calculator is working
> but I got this error :
> 
> CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> at Calculator.render (CalcualtorLogic.js:101)
> at CalcualtorLogic.js:191
> render @ CalcualtorLogic.js:101
> (anonymous) @ CalcualtorLogic.js:191

why did you do that?

**Comment by yaaHosein** [comment]
> > so you mean we can not start this implementation without having the class and id names from before
> > Right?!
> 
> I didn't understand your question

what I want to reach to it that we can not build the render of the OOP calculator if we do not have the html calculator 
because in the OOP calculator 
we have to give class names and id names:

 wrapperDiv.classList.add**("wrapper");**
       
        resultDiv.classList.add**("result");**
        resultDiv.id = **"result";**

**Comment by SMH110** [comment]
> > > so you mean we can not start this implementation without having the class and id names from before
> > > Right?!
> > 
> > 
> > I didn't understand your question
> 
> what I want to reach to it that we can not build the render of the OOP calculator if we do not have the html calculator
> because in the OOP calculator
> we have to give class names and id names:
> 
> wrapperDiv.classList.add**("wrapper");**
> 
> ```
>     resultDiv.classList.add**("result");**
>     resultDiv.id = **"result";**
> ```

still didn't understand what do you mean
what do you mean by OOP? 
do you mean the Calcualtor class?

**Comment by yaaHosein** [comment]
> > I have replaced all resultElement with resultDiv
> > the calculator is working
> > but I got this error :
> > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > at Calculator.render (CalcualtorLogic.js:101)
> > at CalcualtorLogic.js:191
> > render @ CalcualtorLogic.js:101
> > (anonymous) @ CalcualtorLogic.js:191
> 
> why did you do that?

https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697

**Comment by SMH110** [comment]
ab.st

> > > I have replaced all resultElement with resultDiv
> > > the calculator is working
> > > but I got this error :
> > > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > > at Calculator.render (CalcualtorLogic.js:101)
> > > at CalcualtorLogic.js:191
> > > render @ CalcualtorLogic.js:101
> > > (anonymous) @ CalcualtorLogic.js:191
> > 
> > 
> > why did you do that?
> 
> [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697)
can you give the reason?

**Comment by yaaHosein** [comment]
> > > > so you mean we can not start this implementation without having the class and id names from before
> > > > Right?!
> > > 
> > > 
> > > I didn't understand your question
> > 
> > 
> > what I want to reach to it that we can not build the render of the OOP calculator if we do not have the html calculator
> > because in the OOP calculator
> > we have to give class names and id names:
> > wrapperDiv.classList.add**("wrapper");**
> > ```
> >     resultDiv.classList.add**("result");**
> >     resultDiv.id = **"result";**
> > ```
> 
> still didn't understand what do you mean
> what do you mean by OOP?
> do you mean the Calcualtor class?

yes

**Comment by SMH110** [comment]
"what I want to reach to it that we can not build the render of the OOP calculator if we do not have the html calculator
because in the OOP calculator
we have to give class names and id names:"


can you tell me what is the main reason of adding the class name to the elements in the render method?

**Comment by yaaHosein** [comment]
to be able to work with their elements (I mean their parents)

**Comment by yaaHosein** [comment]
> ab.st
> 
> > > > I have replaced all resultElement with resultDiv
> > > > the calculator is working
> > > > but I got this error :
> > > > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > > > at Calculator.render (CalcualtorLogic.js:101)
> > > > at CalcualtorLogic.js:191
> > > > render @ CalcualtorLogic.js:101
> > > > (anonymous) @ CalcualtorLogic.js:191
> > > 
> > > 
> > > why did you do that?
> > 
> > 
> > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697)
> > can you give the reason?

for better understanding :
https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-798967949

**Comment by SMH110** [comment]
> to be able to work with their elements (I mean their parents)

this wrong, 
the main reason of adding class name to the element so they appear styled
for example we added the "result" class name to the result div so the background of the element have this color 
rgba(255, 230, 179, 0.849);

and to have width of 306px

**Comment by SMH110** [comment]
> > ab.st
> > > > > I have replaced all resultElement with resultDiv
> > > > > the calculator is working
> > > > > but I got this error :
> > > > > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > > > > at Calculator.render (CalcualtorLogic.js:101)
> > > > > at CalcualtorLogic.js:191
> > > > > render @ CalcualtorLogic.js:101
> > > > > (anonymous) @ CalcualtorLogic.js:191
> > > > 
> > > > 
> > > > why did you do that?
> > > 
> > > 
> > > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697)
> > > can you give the reason?
> 
> for better understanding :
> [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-798967949)

I had a look at link
I want you to explain what you are trying to do

**Comment by yaaHosein** [comment]
> > to be able to work with their elements (I mean their parents)
> 
> this wrong,
> the main reason of adding class name to the element so they appear styled
> for example we added the "result" class name to the result div so the background of the element have this color
> rgba(255, 230, 179, 0.849);
> 
> and to have width of 306px

Ok
But also the class and id names have come from html file 

And after reading your answer I have another question please 

If do not want to use the css file 

Can I remove line 44 for example?!

**Comment by yaaHosein** [comment]
> > > ab.st
> > > > > > I have replaced all resultElement with resultDiv
> > > > > > the calculator is working
> > > > > > but I got this error :
> > > > > > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > > > > > at Calculator.render (CalcualtorLogic.js:101)
> > > > > > at CalcualtorLogic.js:191
> > > > > > render @ CalcualtorLogic.js:101
> > > > > > (anonymous) @ CalcualtorLogic.js:191
> > > > > 
> > > > > 
> > > > > why did you do that?
> > > > 
> > > > 
> > > > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697)
> > > > can you give the reason?
> > 
> > 
> > for better understanding :
> > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-798967949)
> 
> I had a look at link
> I want you to explain what you are trying to do

Trying to find the difference between result Div and result element 
Because these two are causing a lot of confusion for me 
And for understanding why we have them both 
Why not only one of them

**Comment by SMH110** [comment]
> > > > ab.st
> > > > > > > I have replaced all resultElement with resultDiv
> > > > > > > the calculator is working
> > > > > > > but I got this error :
> > > > > > > CalcualtorLogic.js:101 Uncaught TypeError: Cannot read property 'appendChild' of null
> > > > > > > at Calculator.render (CalcualtorLogic.js:101)
> > > > > > > at CalcualtorLogic.js:191
> > > > > > > render @ CalcualtorLogic.js:101
> > > > > > > (anonymous) @ CalcualtorLogic.js:191
> > > > > > 
> > > > > > 
> > > > > > why did you do that?
> > > > > 
> > > > > 
> > > > > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-787449697)
> > > > > can you give the reason?
> > > 
> > > 
> > > for better understanding :
> > > [#23 (comment)](https://github.com/yaaHosein/Calculator1/pull/23#issuecomment-798967949)
> > 
> > 
> > I had a look at link
> > I want you to explain what you are trying to do
> 
> Trying to find the difference between result Div and result element
> Because these two are causing a lot of confusion for me
> And for understanding why we have them both
> Why not only one of them


This why I asked you add a comment for everything 
because now I have to explain this again to you, despite I explained this before many time

`resultDiv` is constant that's hold reference of the actual HTML div element when we created it like this

```js
const resultDiv = document.createElement('div');
```
and because we will need to this element (the actual HTML element) later when we need to update the result div 
because of that, we save it as property on the class 

because you only have access to the constant `resultDiv` inside the `render` method 
outside the `render` method you can use `resultDiv` because it's not defined 

have a look at this code where we really use the `resultDiv` 
```js
updateResultElement() {
        this.resultElement.textContent = this.currentResult;
    }
```
here we used the `resultDiv` but not by typing `resultDiv` but by doing `this.resultElement` 

in the `render` method we saved the `resultDiv` which is a HTML div in `resultElement` property 


if this really consusing you and you want to use only of them then you can do  it like this


```js

render() {
        const wrapperDiv = document.createElement('div');
        wrapperDiv.classList.add("wrapper");
        this.resultElement  = document.createElement('div');
        this.resultElement .classList.add("result");
        this.resultElement .id = "result";
        this.resultElement .appendChild(
            document.createTextNode(this.currentResult)
        );
        wrapperDiv.appendChild(this.resultElement );

 /// ... 
```


now you tell me which is easier?

**Comment by yaaHosein** [comment]
Thanks a lot 
First I have to read this comment again and again until I could understand as much it is possible

**Comment by SMH110** [comment]
> Thanks a lot
> First I have to read this comment again and again until I could understand as much it is possible

can you tell me what you didn't understand, will try to help you

**Comment by yaaHosein** [comment]
Thank you so much 
First I want to try myself

**Comment by yaaHosein** [comment]
> ```js
> this.resultElement );
> 
>  

this.resultElement is very better 

I was thinking to change 
resultElement to become 
resultDiv_
Or the opposed 

But your code also fine

Regarding to the error in line 101 
Which I mistakenly told you it appeared after changing resultDiv to be resultElemet

I found it was because of instant class
Because I was not changed the css class selector 

After updating it to be the host id name the problem is solved

**Comment by SMH110** [comment]
The main reason of this Pull request is to add history information 
so in the new class we need to access the wrapper element and add another div to it 
this new div will be hold the content of the history information

The best way to do that is to define a render method in the new class 
note, the new class inherited the render method of its parent.
this means when you define render method in  the new class, you will overwrite the render method of its parent 

see this code
```js

const developedCalculator = new DevelopedCalculator("#hostElement_");
developedCalculator.render()
```

currently, you're calling the `redner` method on instance of DevelopedCalculator class
this render method is defined in the parent of `DevelopedCalculator` class 

if you add render method to `DevelopedCalculator` then render method of the parent class will become inactive 
in other words, it will not be called by JS .

saying that, you still have access to the parent render method 
the `super` keyword give you access to the parent class.

so in the new render method, call the render method of the parent 

`super.render()`

Please don't touch the `CalcualtorLogic.js`
If I see any changes in the `CalcualtorLogic.js` I will ask you close this PR and revert all your changes and start from scratch again
